src/Agent.Worker/Maintenance/MaintenanceJobExtension.cs

```diff
@@ -0,0 +1,74 @@
+using Microsoft.VisualStudio.Services.Agent.Util;
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Maintenance
+{
+    public interface IMaintenanceServiceProvider : IExtension
+    {
+        string MaintenanceDescription { get; }
+        void RunMaintenanceOperation(IExecutionContext context);
+    }
+
+    public sealed class MaintenanceJobExtension : AgentService, IJobExtension
+    {
+        public Type ExtensionType => typeof(IJobExtension);
+        public string HostType => "poolmaintenance";
+        public IStep PrepareStep { get; private set; }
+        public IStep FinallyStep { get; private set; }
+
+        public MaintenanceJobExtension()
+        {
+            PrepareStep = new JobExtensionRunner(
```

Comment:
Depending on which PR goes first, need to see what impact Condition has on this (references `Agent.JobStatus == succeeded`) for normal condition.